ledger/api-server-damlonx/reference-v2/src/test/suite/scala/com/daml/ledger/api/server/damlonx/reference/v2/IndexerIT.scala

```diff
@@ -0,0 +1,280 @@
+        loggerFactory.logs(classOf[RecoveringIndexer]) should contain theSameElementsAs Seq(
+          Level.INFO -> "Starting Indexer Server",
+          Level.INFO -> "Started Indexer Server",
+          Level.INFO -> "Stopping Indexer Server",
+          Level.ERROR -> "Error while running indexer, restart scheduled after 10 seconds",
+          Level.INFO -> "The Indexer Server was stopped, so not restarting",
+          Level.INFO -> "Stopped Indexer Server",
+        )
+      }
+    }
+  }
+
+  private def initializeEverything(
+      newParticipantState: (ParticipantId, LedgerString) => ParticipantState,
+      restartDelay: FiniteDuration = 100.millis,
+  ): (ParticipantState, Resource[Unit], Resource[LedgerDao]) = {
+    val id = UUID.randomUUID()
+    val participantId: ParticipantId = LedgerString.assertFromString(s"participant-$id")
+    val ledgerId = LedgerString.assertFromString(s"ledger-$id")
+    val participantState = newParticipantState(participantId, ledgerId)
+    val jdbcUrl =
+      s"jdbc:h2:mem:${getClass.getSimpleName.toLowerCase()}-$id;db_close_delay=-1;db_close_on_exit=false"
+    val serverOwner = new StandaloneIndexerServer(
+      participantState,
+      IndexerConfig(
+        participantId,
+        jdbcUrl,
+        startupMode = IndexerStartupMode.MigrateAndStart,
+        restartDelay = restartDelay,
+      ),
+      loggerFactory,
+      SharedMetricRegistries.getOrCreate(s"${getClass.getSimpleName}-server"),
+    )
+    val ledgerDaoOwner =
+      JdbcLedgerDao.owner(
+        jdbcUrl,
+        loggerFactory,
+        SharedMetricRegistries.getOrCreate(s"${getClass.getSimpleName}-ledger-dao"),
+        ExecutionContext.global,
+      )
+    val server = serverOwner.acquire()
+    val ledgerDao = ledgerDaoOwner.acquire()
+    (participantState, server, ledgerDao)
+  }
+}
+
+object IndexerIT {
+  private type ParticipantState = ReadService with WriteService with AutoCloseable
+
+  private val eventually = RetryStrategy.exponentialBackoff(10, 10.millis)
+
+  private val loggerFactory = TestNamedLoggerFactory(classOf[IndexerIT])
+
+  // This class inserts a failure after each state update to force the RecoveringIndexer to restart.
+  private class ParticipantStateWhichFailsOften(delegate: ParticipantState)
+      extends DelegatingParticipantState(delegate) {
+    private var lastFailure: Option[Offset] = None
+
+    override def stateUpdates(beginAfter: Option[Offset]): Source[(Offset, Update), NotUsed] =
+      super
+        .stateUpdates(beginAfter)
+        .flatMapConcat {
+          case value @ (offset, _) =>
+            if (lastFailure.isEmpty || lastFailure.get < offset) {
+              lastFailure = Some(offset)
+              Source.single(value).concat(Source.failed(new StateUpdatesFailedException))
+            } else {
+              Source.single(value)
+            }
+        }
+  }
+
+  private class DelegatingParticipantState(delegate: ParticipantState)
```

Comment:
You may want to consider using a Mockito spy here stubbing only the `stateUpdates` call to avoid having to maintain this class definition as well as we evolve participant state interfaces:
https://github.com/mockito/mockito/wiki/Using-Spies-(and-Fakes)

Reply:
Good idea. Done.